cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/DefaultSystemAppTaskContext.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Default implementation for {@link io.cdap.cdap.api.service.worker.SystemAppTaskContext}
+ */
+public class DefaultSystemAppTaskContext extends AbstractServiceDiscoverer implements SystemAppTaskContext {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultSystemAppTaskContext.class);
+
+  private final PreferencesFetcher preferencesFetcher;
+  private final CConfiguration cConf;
+  private final PluginFinder pluginFinder;
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final SecureStore secureStore;
+  private final String serviceName;
+  private final RetryStrategy retryStrategy;
+  private final ArtifactManager artifactManager;
+  private final PluginInstantiator instantiator;
+  private final File pluginsDir;
+  private final io.cdap.cdap.proto.id.ArtifactId protoArtifactId;
+
+  DefaultSystemAppTaskContext(CConfiguration cConf, PreferencesFetcher preferencesFetcher, PluginFinder pluginFinder,
+                              DiscoveryServiceClient discoveryServiceClient, SecureStore secureStore,
+                              String artifactNameSpace, ArtifactId artifactId, ClassLoader artifactClassLoader,
+                              ArtifactManagerFactory artifactManagerFactory, String serviceName) {
+    super(artifactNameSpace);
+    this.cConf = cConf;
+    this.serviceName = serviceName;
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Retry.SERVICE_PREFIX);
+    this.preferencesFetcher = preferencesFetcher;
+    this.pluginFinder = pluginFinder;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.secureStore = secureStore;
+    this.artifactManager = artifactManagerFactory.create(new NamespaceId(artifactNameSpace), retryStrategy);
+    this.pluginsDir = createTempFolder();
+    instantiator = new PluginInstantiator(cConf, artifactClassLoader, pluginsDir);
+    protoArtifactId = new io.cdap.cdap.proto.id.ArtifactId(artifactNameSpace, artifactId.getName(),
+                                                           artifactId.getVersion().getVersion());
+  }
+
+  private File createTempFolder() {
+    File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    try {
+      return Files.createTempDirectory(tmpDir.toPath(), "plugins").toFile();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Map<String, String> getPreferencesForNamespace(String namespace, boolean resolved) throws IOException {
+    try {
+      return preferencesFetcher.get(new NamespaceId(namespace), resolved).getProperties();
```

Comment:
Seems like this should get wrapped with Retries. You might need to fix `BasicSystemHttpServiceContext` as well

Reply:
Done.